### PR TITLE
Use app label as app-change label for new apps

### LINKED ISSUE #646: App name cannot have more than 63 characters
**What steps did you take:**
Created an app with the app name having more than 63 characters.

**What happened:**
App creation failed because the app-change ConfigMaps uses the app name as a label.
```
kapp: Error: Creating app change:
  ConfigMap "byoi-2213-photon-ns-antrea-11ns1-c1-guest-cluster-auth-serzfmpw" is invalid: metadata.labels:
    Invalid value: "byoi-2213-photon-ns-antrea-11ns1-c1-guest-cluster-auth-service-ctrl": must be no more than 63 characters
```

**What did you expect:**
App to be created successfully.

**Anything else you would like to add:**
This causes issues for App CRs and PackageInstall CRs from kapp-controller because a kapp app is created using the name of app/package install.

**Environment:**
- kapp version (use `kapp --version`): kapp controller version is `v0.41.2`
- OS (e.g. from `/etc/os-release`): `Photon`
- Kubernetes version (use `kubectl version`): `1.22.13`

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by praveenrewar**
#### What this PR does / why we need it:
- Introduce a new label `kapp.k14s.io/app-change-app-label` for app changes which uses the app label as value
- Use the old label to list and delete app-changes created for exiting apps, but also add this new label to them.
- Use the new label to list and delete app-changes created for new apps, but add the old label if length of app name is less than the maximum allowed length of a label key to keep the app-change backward compatible

#### Which issue(s) this PR fixes:

Fixes #646 

#### Does this PR introduce a user-facing change?

```release-note
Breaking change!!
App label generated by kapp will be used as app-change label instead of the app name for all the new apps.
- Using app name as a label value in app changes puts a limit on the number of characters a app name can have
- If an app deployed with new version of kapp (and having more than 63 characters) is used with an old version of kapp, then the old version won't be able to detect the app-changes created by the new version.
```

#### Additional Notes for your reviewer:
In future, after x number of releases, we can stop using the legacy label and start using the new label for existing apps as well. Since all the app-changes created for them would most probably have both the labels, it won't result in any loss of app-changes.

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kapp/app/recorded_app.go [hunk 1]
```diff
@@ -26,13 +26,16 @@ const (
 	KappIsConfigmapMigratedAnnotationKey   = "kapp.k14s.io/is-configmap-migrated"
 	KappIsConfigmapMigratedAnnotationValue = ""
 	AppSuffix                              = ".apps.k14s.io"
+	KappIsAppLabelAsAppChangeLabelKey      = "kapp.k14s.io/is-app-label-app-change-label"
```
Comment (100mik) [inline pkg/kapp/app/recorded_app.go]: "kapp.k14s.io/uses-app-label" maybe? A bit of a tongue twister right now

File: pkg/kapp/app/recorded_app.go [hunk 2]
```diff
@@ -585,6 +604,19 @@ func (a *RecordedApp) update(doFunc func(*Meta)) error {
 	return nil
 }
 
+func (a *RecordedApp) changeLabelValue() (string, error) {
```
Comment (100mik) [inline pkg/kapp/app/recorded_app.go]: nit: `appChangeLabelValue` feels more comprehensive to me.
Same for all the places where we use `changeLabelValue` 🤔

File: pkg/kapp/app/recorded_app.go (hunk 1 above)
Comment (100mik) [reply pkg/kapp/app/recorded_app.go]: Or I guess `changes-use-app-id` is more appropriate

File: pkg/kapp/app/recorded_app.go [hunk 3]
```diff
@@ -26,13 +26,16 @@ const (
...
 type RecordedApp struct {
-	name              string
-	nsName            string
-	isMigrated        bool
-	creationTimestamp time.Time
+	name                     string
+	nsName                   string
+	isMigrated               bool
+	creationTimestamp        time.Time
+	isAppLabelAppChangeLabel bool
```
Comment (100mik) [inline pkg/kapp/app/recorded_app.go]: `isAppLabelAppChangeLabel`  => `UsesAppID` ?

File: pkg/kapp/app/recorded_app.go [hunk 4]
```diff
@@ -146,6 +150,7 @@ func (a *RecordedApp) CreateOrUpdate(labels map[string]string, isDiffRun bool) (
 	}
 
 	if foundNonMigratedApp {
+		_, a.isAppLabelAppChangeLabel = app.Annotations[KappIsAppLabelAsAppChangeLabelKey]
```
Comment (100mik) [inline pkg/kapp/app/recorded_app.go]: Since we are doing this for both migrated and non-migrated apps, why don't we populate this value before we make checks for migration?
I do not see the "name migration" and "label migration" being coupled in any way, am I missing something?

File: pkg/kapp/app/recorded_app.go [hunk 5]
```diff
@@ -174,6 +179,10 @@ func (a *RecordedApp) create(labels map[string]string, isDiffRun bool) error {
 			Labels: map[string]string{
 				KappIsAppLabelKey: kappIsAppLabelValue,
 			},
+			// Use app label as app change label for all new apps
+			Annotations: map[string]string{
+				KappIsAppLabelAsAppChangeLabelKey: "",
```
Comment (100mik) [inline pkg/kapp/app/recorded_app.go]: `KappIsAppLabelAsAppChangeLabelKey` => `KappChangesUseAppID` ?
I get that we are trying to say that the AppChange label uses the label that app resources have. I just think it can be simpler 🤔 
Any other ideas?

File: pkg/kapp/app/recorded_app.go [hunk 6]
```diff
@@ -415,6 +427,9 @@ func (a *RecordedApp) Rename(newName string, newNamespace string) error {
 func (a *RecordedApp) renameConfigMap(app *corev1.ConfigMap, name, ns string) error {
 	oldName := app.Name
 
+	// Use app label as app change label for new, renamed or migrated apps
```
Comment (100mik) [inline pkg/kapp/app/recorded_app.go]: How about apps that have been migrated, but use app label name instead of ID right now. It is not super clear to me how we are handling those cases 🤔 
1. Will these only be renamed _if_ the user renames the app?
2. Does renaming the app to the same name rename CMs?

**Review by 100mik — COMMENTED** [review]
A few thoughts and I am thinking about whether migrated apps that use the label name today is being considered.

File: pkg/kapp/app/recorded_app.go (hunk 1 above)
Comment (praveenrewar) [reply pkg/kapp/app/recorded_app.go]: > "kapp.k14s.io/uses-app-label" maybe? A bit of a tongue twister right now

Yeah, I agree, but "uses-app-label" might not be clearly indicating that this is for app changes.

> Or I guess changes-use-app-id is more appropriate

Hmm, I am not sure if we ever use `app-id` as a term for the app label value 🤔

File: pkg/kapp/app/recorded_app.go (hunk 2 above)
Comment (praveenrewar) [reply pkg/kapp/app/recorded_app.go]: Sure, I can make the change. I think I kept it `changeLabelValue` only because the method is on RecordedApp, so it will become `app.appChangeLabelValue` in some places.

File: pkg/kapp/app/recorded_app.go (hunk 1 above)
Comment (praveenrewar) [reply pkg/kapp/app/recorded_app.go]: Maybe "app-changes-use-app-label"?

File: pkg/kapp/app/recorded_app.go (hunk 2 above)
Comment (100mik) [reply pkg/kapp/app/recorded_app.go]: I would prefer the latter because "app" and "appChange" are two separate values from my perspective 🤔

File: pkg/kapp/app/recorded_app.go (hunk 4 above)
Comment (praveenrewar) [reply pkg/kapp/app/recorded_app.go]: Yeah, I was thinking of keeping it outside as well, the only reason I didn't do it is because if `!foundNonMigratedApp` then app would also be empty, but that should be okay, so I am totally fine with making this change.

File: pkg/kapp/app/recorded_app.go (hunk 6 above)
Comment (praveenrewar) [reply pkg/kapp/app/recorded_app.go]: > but use app label name instead of ID

You mean app name instead of ID? If the app is _already_ migrated before deploying with newer version, then there won't be any change, but if migration happens with the newer version of kapp, then it will start using the app label instead of the name.

> Will these only be renamed if the user renames the app?

Yep. So, we will start using app labels instead of app name for either new apps or when a new meta configmap is being created (i.e. the existing app changes are already being abandoned), which happens in case of rename, migration and using `--prev-app` flag with deploy. I have mentioned all the scenarios in the table.

> Does renaming the app to the same name rename CMs?

We first create the new configmap and then delete the existing one, so if the user provides the same name as `--new-name` to `kapp rename` command, then it will through an error saying the configmap already exists.

File: pkg/kapp/app/recorded_app.go (hunk 2 above)
Comment (praveenrewar) [reply pkg/kapp/app/recorded_app.go]: Makes sense.

File: pkg/kapp/app/recorded_app.go (hunk 1 above)
Comment (rohitagg2020) [reply pkg/kapp/app/recorded_app.go]: `app-change-uses-app-label`?

Comment (100mik) [reply pkg/kapp/app/recorded_app.go]: That feels more comprehensive ^
Lets do this

File: pkg/kapp/app/recorded_app.go (hunk 3 above)
Comment (100mik) [reply pkg/kapp/app/recorded_app.go]: `AppChangesUseAppLabel` in line with what we discussed earlier

**Comment by 100mik** [comment]
I see that you called out that we won't be any change for existing + migrated apps. That checks out.
The only sore point I have is this means that we will never entirely stop using the old way.
The changes seem to be in line with the behaviour that you are calling out.

What I was wondering is, since app-changes are something we garbage collect. Would it make sense to do this at an app-change level?
1. All new app-changes have app label
2. We query for both labels, _unless_ the `app-changes-use-app-label` annotation is set.
3. We set the annotation when we get an empty list when looking for app-changes with the old label

So the **huge** downside is we make one more call till the app is migrated. And from a kapp-controller perspective. For stable packages that do not have diffs, this might be forever till they are reinstalled.

I am trying to figure out how we can _potentially_ move to using *just* the new label. The path we are taking right now feels like we will be handling both cases for a long time since we cannot definitely said that no `migrated` or `existing` apps exist on the users cluster. And we end up leaving orphaned config maps when we do decide to move to the new one altogether.

Since, this is a case that is consciously not being handled in the PR. We can however track how we can "taper off"  using old labels as a separate issue since this fix might be needed sooner rather than later.

TL;DR: The path to us not supporting the old label at all in the future is not clear. But we might track that separately and improve on the cases that we do not handle.
@praveenrewar thoughts?

**Comment by praveenrewar** [comment]
@100mik Thanks for bringing this up. I was planning to take that up separately, but now I think that we should at least decide how we migrate the app-changes for existing apps before making this change. 

How about we add one more label to app-changes (not label value, but new label itself). For new apps we will use the new label itself, for old apps we will use the old label but also add the new label to the app changes. After some releases we will switch to using the new label only.

New app-changes for existing apps:
```
labels:
    kapp.k14s.io/app-change-app: foo
    kapp.k14s.io/app-change-app-label: 101920101920
    kapp.k14s.io/is-app-change: ""
```

New app-changes for new apps:
```
labels:
    kapp.k14s.io/app-change-app-label: 101920101920
    kapp.k14s.io/is-app-change: ""
```

Does this make sense? I will see how these changes looks from code perspective.

File: pkg/kapp/app/recorded_app.go (hunk 4 above)
Comment (100mik) [reply pkg/kapp/app/recorded_app.go]: Maybe do it after a check on `app` ? That would make sense I think

File: pkg/kapp/app/recorded_app.go (hunk 5 above)
Comment (100mik) [reply pkg/kapp/app/recorded_app.go]: We can change this in line with the others !

File: pkg/kapp/app/recorded_app.go (hunk 4 above)
Comment (100mik) [reply pkg/kapp/app/recorded_app.go]: The function `RenamePrevApp` seems to have similar logic and should also reflect this change?

**Comment by 100mik** [comment]
Yeah, I think adding both labels now does give us a part forward. We should do that if it is feasible 🤔

File: pkg/kapp/app/recorded_app.go (hunk 4 above)
Comment (praveenrewar) [reply pkg/kapp/app/recorded_app.go]: > Maybe do it after a check on app ?

We return from the check on app.

**Review by 100mik — APPROVED** [review]
The changes look good to me, adding both labels leave more room for us to plan out the migration!
I was wondering if there is a straightforward way of putting together tests for this, but I believe we do not have any form of tests involving older versions of `kapp`. So I am happy with these changes as long as we have verified the tabulated test cases. Let me know if you need a hand with the verification!

**Review by 100mik — APPROVED** [review]
The additional condition makes sense!

File: pkg/kapp/app/recorded_app_changes.go [hunk 7]
```diff
@@ -17,22 +17,24 @@ import (
...
 
-	appLabelValue string
+	appChangeUsesAppLabel bool
 
 	coreClient kubernetes.Interface
 }
 
-func NewRecordedAppChanges(nsName, appName, appLabelValue string, coreClient kubernetes.Interface) RecordedAppChanges {
-	return RecordedAppChanges{nsName, appName, appLabelValue, coreClient}
+func NewRecordedAppChanges(nsName, appName, changeLabelValue string, appChangeUsesAppLabel bool, coreClient kubernetes.Interface) RecordedAppChanges {
```
Comment (rohitagg2020) [inline pkg/kapp/app/recorded_app_changes.go:36]:
```suggestion
func NewRecordedAppChanges(nsName, appName, appChangeLabelValue string, appChangeUsesAppLabel bool, coreClient kubernetes.Interface) RecordedAppChanges {
```

**Review by rohitagg2020 — APPROVED** [review]

File: pkg/kapp/app/recorded_app_changes.go (hunk 7 above)
Comment (kumaritanushree) [reply pkg/kapp/app/recorded_app_changes.go:36]: +1

Comment (praveenrewar) [reply pkg/kapp/app/recorded_app_changes.go:36]: (I kept it `changeLabelValue` to be consistent with `changeLabelKey`)
Do you want me to make this change before merging?
